daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/KeyStateMachine.scala

```diff
@@ -0,0 +1,456 @@
+          }
+          // Since keys is defined only where keyInputs is defined, we don't need to update keyInputs.
+          Either.cond(keyMapping == KeyActive(cid), next, InconsistentKeys(gk))
+      }
+    }
+
+    def handleLeaf(leaf: Node.LeafOnlyAction): Either[KeyInputError, State] =
+      leaf match {
+        case create: Node.Create => handleCreate(create)
+        case fetch: Node.Fetch => handleFetch(fetch)
+        case lookup: Node.LookupByKey => handleLookup(lookup)
+      }
+
+    /** To be called when interpretation enters a try block or iteration enters a Rollback node
+      * Must be matched by [[endRollback]] or [[dropRollback]].
+      */
+    def beginRollback(): State = {
+      this.copy(rollbackStack = activeState +: rollbackStack)
+    }
+
+    /** To be called when interpretation does insert a Rollback node or iteration leaves a Rollback node.
+      * Must be matched by a [[beginRollback]].
+      */
+    def endRollback(): State = rollbackStack match {
+      case Nil => throw new IllegalStateException("Not inside a rollback scope")
+      case headState :: tailStack => this.copy(activeState = headState, rollbackStack = tailStack)
+    }
+
+    /** To be called if interpretation notices that a try block did not lead to a Rollback node
+      * Must be matched by a [[beginRollback]].
+      */
+    def dropRollback(): State = rollbackStack match {
+      case Nil => throw new IllegalStateException("Not inside a rollback scope")
+      case _ :: tailStack => this.copy(rollbackStack = tailStack)
+    }
+
+    private def withinRollbackScope: Boolean = rollbackStack.nonEmpty
+
+    /** Let `resolver` be a [[KeyResolver]] that can be used during interpretation to obtain a transaction `tx`.
+      * Let `this` state be the result of iterating over `tx` up to a node `n` exclusive using `resolver`.
+      * Let `substate` be the state obtained after fully iterating over the subtree rooted at `n` starting from [[State.empty]],
+      * using the resolver [[projectKeyResolver]](`resolver`).
+      * Then, the returned state is the same as if
+      * the iteration over tx continued from `this` state through the whole subtree rooted at `n` using `resolver`.
+      *
+      * If the iteration over the subtree rooted at `n` from `this` using the projected resolver fails,
+      * then so does advancing `this` using `substate` with the same error.
+      * If iteration of `n` from both `this` and [[State.empty]] fail,
+      * the error kind ([[DuplicateContractKey]] vs [[InconsistentKeys]]) may differ.
+      * For example, for `tx = [ Create c0 (key = k), Exe c1 [ Create c2 (key = k), LookupByKey k -> None ] ]`
+      * with `n` being the `Exe c1` node, traversing all of `tx` errors with [[DuplicateContractKey]] on `Create c2`
+      * whereas traversing the `n` subtree from [[State.empty]] errors with [[InconsistentKeys]] on `LookupByKey k`.
+      */
+    // TODO(#9386) add test cases
```

Comment:
If you drop this, I think I’d be fine merging this without additional tests since the rest is already tested indirectly and then we can add tests as a follow-up PR. Might make at least reviewing a bit easier. wdyt?

Reply:
Good idea. Removed for now.